geode-cq/src/main/java/org/apache/geode/cache/query/cq/internal/ServerCQImpl.java

```diff
@@ -63,13 +62,13 @@
    * NOTE: In case of RR this map is populated and used as intended. In case of PR this map will not
    * be populated. If executeCQ happens after update operations this map will remain empty.
    */
-  private volatile HashMap<Object, Object> cqResultKeys;
+  private volatile ConcurrentMap<Object, Object> cqResultKeys;
 
   /**
    * This maintains the keys that are destroyed while the Results Cache is getting constructed. This
    * avoids any keys that are destroyed (after query execution) but is still part of the CQs result.
    */
-  private HashSet<Object> destroysWhileCqResultsInProgress;
+  private CopyOnWriteHashSet<Object> destroysWhileCqResultsInProgress;
```

Comment:
Does this have a heavy impact on creating a lot more garbage?

Reply:
Thinking of the case where we have a lot of cq's

Reply:
Good question... I haven't seen a major impact on the memory footprint during the performance tests. Whenever we execute an operation on the `CopyOnWriteHashSet` the "old instance" is ready to be garbage collected, though, so I wouldn't expect major issues.
Am I missing something?. If this becomes a blocker for approving this `PR`, please let me know and I'll see what I can do.